### PR TITLE
Allow to customise directory and notebook icon

### PR DESCRIPTION
**Pull request by linlol**
## References

for issue 15741 https://github.com/jupyterlab/jupyterlab/issues/15741

## Code changes

as I proposed previously

currently for notebook and directory, Jupyter would ignore user-specification.

In my change, I would force lookup file type with respect to suffix/pattern, and check if the content type coincides with notebook/directory

as discussed with @krassowski and @JasonWeill , this shall be part of 4.2

## User-facing changes

I believe no impact to current user (if no bug)

after implemented, user can customise icon via writing lab extensions

## Backwards-incompatible changes

I guess no backwards incompatibility breaking since it is small change

### CODE REVIEW

**Comment by linlol** [comment]
@krassowski do you know how can I check the detail of those UI integration test failure? those regression failure looks irrelevant and I can see similar issue happens frequently in other PR

**Comment by krassowski** [comment]
Yes, they look unrelated, I restarted them.

**Comment by linlol** [comment]
> Yes, they look unrelated, I restarted them.

Thanks @krassowski for restarting it, may I know are those UI regression failure regular as I can see a few other pull requests with similar failures.

Also, (sorry that I am new to the development of Jupyter itself) honestly I don't know how to have a quick test, is there any approach to build Jupyter and link the JS target to asset like what we did in lab extension development

**Comment by linlol** [comment]
looks like one test passed after retry and the other still failing, can someone help re-trigger the failed one

**Comment by linlol** [comment]
Hello @krassowski, @JasonWeill. Sorry that I was AFK for a few personal issue.

Looks like my change works

My sample change:
```
import {juliaIcon} from '@jupyterlab/ui-components';
    app.docRegistry.addFileType({
      name: 'node_module',
      contentType: 'directory',
      icon: juliaIcon,
      pattern: '^node_modules$'
    })
```
As you can see below, looks like icon of node_modules is changed to JuliaIcon
image

Since looks like it does not break any specific ut, let's try to fix the pipeline and merge it?

**Comment by linlol** [comment]
test keeps failing with words like
pluggy._manager.PluginValidationError: Plugin 'check-links' for hook 'pytest_collect_file'
hookimpl definition: pytest_collect_file(path: 'Any', parent: 'pytest.Collector') -> 'CheckLinks | None'
Argument(s) {'path'} are declared in the hookimpl but can not be found in the hookspec

Is there something broken by my change？

**Comment by krassowski** [comment]
@linlol this appears a recent change in a dependency which started showing up on CI today. We are investigating in https://github.com/jupyterlab/jupyterlab/issues/15908

**Comment by linlol** [comment]
> @linlol this appears a recent change in a dependency which started showing up on CI today. We are investigating in #15908

Thanks @krassowski! would merge main after issue resolved

**Comment by linlol** [comment]
somehow I got this error...

is this something like version auto-bump? @krassowski 


Made integrity changes!
Please commit the changes by running:
git commit -a -m "Package integrity updates"
4.1.3
4.1.3
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.8/x64/bin/jupyter-releaser", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.8/x64/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.8/x64/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/runner/work/_actions/jupyter-server/jupyter_releaser/v2/jupyter_releaser/cli.py", line 121, in invoke
    super().invoke(ctx)
  File "/opt/hostedtoolcache/Python/3.11.8/x64/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.8/x64/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.8/x64/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/_actions/jupyter-server/jupyter_releaser/v2/jupyter_releaser/cli.py", line 382, in bump_version
    lib.bump_version(version_spec, version_cmd, changelog_path, tag_format)
  File "/home/runner/work/_actions/jupyter-server/jupyter_releaser/v2/jupyter_releaser/lib.py", line 43, in bump_version
    raise ValueError(msg)
ValueError: Tag v4.1.3 already exists! To delete run: `git push --delete origin {tag_name}`
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/runner/work/_actions/jupyter-server/jupyter_releaser/v2/jupyter_releaser/actions/prep_release.py", line 25, in <module>
    run_action("jupyter-releaser bump-version")
  File "/home/runner/work/_actions/jupyter-server/jupyter_releaser/v2/jupyter_releaser/actions/common.py", line 25, in run_action
    _run(target, *args, **kwargs)
  File "/home/runner/work/_actions/jupyter-server/jupyter_releaser/v2/jupyter_releaser/util.py", line 94, in run
    raise e
  File "/home/runner/work/_actions/jupyter-server/jupyter_releaser/v2/jupyter_releaser/util.py", line 86, in run
    process = tee(cmd, **kwargs)
              ^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/_actions/jupyter-server/jupyter_releaser/v2/jupyter_releaser/tee.py", line 159, in run
    raise subprocess.CalledProcessError(
subprocess.CalledProcessError: Command 'jupyter-releaser bump-version' returned non-zero exit status 1.

**Comment by krassowski** [comment]
@linlol just an unlucky timing - the CI got confused as it started before 4.1.3 got released before it was done it was already released; I restarted the test it should pass now.

**Comment by linlol** [comment]
> @linlol just an unlucky timing - the CI got confused as it started before 4.1.3 got released before it was done it was already released; I restarted the test it should pass now.

@krassowski lol it happens

**Comment by linlol** [comment]
hi @krassowski can you retrigger it again? looks like during check-link, github raises a series of 429 too many request, I guess it is due to many similar check-link checks running at the same moment?

Also we can notice UI regression failure, I guess it is expected, please correct me if I am wrong

regards

**Comment by linlol** [comment]
@krassowski similar 429 for github request for check link failure,

and do you know what does this mean

image

how can I check which test failed here?

**Comment by krassowski** [comment]
That job timed out (it has a timeout of 30 minutes, it is not clear why it got stalled but could be a runner issue) - I restarted it.

**Comment by linlol** [comment]
@krassowski I see, any suggestion for check-link failure. Looks like it failed 3 or 4 times in a row, shall we make any change or just retry

**Comment by linlol** [comment]
@krassowski Hi, it finally passed all checks, could you help check if it is ok to merge it?

**Comment by krassowski** [comment]
Thanks! I think we still should add a test to ensure that this does not regress in the future when someone will be optimizing this function. The test should go in:

https://github.com/jupyterlab/jupyterlab/blob/b6253d94454818b3feefbbe7a095b0da3286cfe9/packages/docregistry/test/registry.spec.ts#L682-L736

**Comment by linlol** [comment]
> Thanks! I think we still should add a test to ensure that this does not regress in the future when someone will be optimizing this function. The test should go in:
> 
> https://github.com/jupyterlab/jupyterlab/blob/b6253d94454818b3feefbbe7a095b0da3286cfe9/packages/docregistry/test/registry.spec.ts#L682-L736

@krassowski sure, I would handle the ut this weekend. Btw, is it correct that no UI regression requested

**Comment by linlol** [comment]
hello @krassowski, added unit test for notebook/directory's customised FileType support and all [Linux JS Tests](https://github.com/jupyterlab/jupyterlab/actions/workflows/linuxjs-tests.yml) passed.

File: packages/docregistry/test/registry.spec.ts [hunk 1]
```diff
@@ -700,6 +700,42 @@ describe('docregistry/registry', () => {
...
+        const regularDirectoryFt = registry.getFileTypeForModel({
+          path: '/foo',
+          type: 'directory'
+        });
+        expect(regularDirectoryFt.name).toBe('directory');
+        const nodeNoduleFt = registry.getFileTypeForModel({
+          path: '/foo/node_modules',
+          type: 'directory'
+        });
+        expect(nodeNoduleFt.name).toBe('node_module');
```
Comment (krassowski) [inline packages/docregistry/test/registry.spec.ts:718]: Just for clarity for a future reader:

```suggestion
        const nodeModuleFt = registry.getFileTypeForModel({
          path: '/foo/node_modules',
          type: 'directory'
        });
        expect(nodeModuleFt.name).toBe('node_module');
```

File: packages/docregistry/test/registry.spec.ts [hunk 2]
```diff
@@ -700,6 +700,42 @@ describe('docregistry/registry', () => {
...
+        const regularDirectoryFt = registry.getFileTypeForModel({
+          name: 'foo.ipynb',
+          type: 'notebook'
+        });
+        expect(regularDirectoryFt.name).toBe('notebook');
+        const nodeNoduleFt = registry.getFileTypeForModel({
+          path: 'test.ipynb',
+          type: 'notebook'
+        });
+        expect(nodeNoduleFt.name).toBe('test_ipynb');
```
Comment (krassowski) [inline packages/docregistry/test/registry.spec.ts:736]:
```suggestion
        const customNotebookType = registry.getFileTypeForModel({
          path: 'test.ipynb',
          type: 'notebook'
        });
        expect(customNotebookType.name).toBe('test_ipynb');
```

File: packages/docregistry/test/registry.spec.ts [hunk 3]
```diff
@@ -700,6 +700,42 @@ describe('docregistry/registry', () => {
...
+        });
+        expect(regularDirectoryFt.name).toBe('directory');
+        const nodeNoduleFt = registry.getFileTypeForModel({
+          path: '/foo/node_modules',
+          type: 'directory'
+        });
+        expect(nodeNoduleFt.name).toBe('node_module');
+      });
+
+      it('allow customise filetype for notebook', () => {
```
Comment (krassowski) [inline packages/docregistry/test/registry.spec.ts:721]:
```suggestion
      it('should allow to customise filetype for notebook', () => {
```

File: packages/docregistry/test/registry.spec.ts [hunk 4]
```diff
@@ -700,6 +700,42 @@ describe('docregistry/registry', () => {
         expect(ft.name).toBe('notebook');
       });
 
+      it('allow customise filetype for directory', () => {
```
Comment (krassowski) [inline packages/docregistry/test/registry.spec.ts:703]:
```suggestion
      it('should allow to customise filetype for directory', () => {
```

File: packages/docregistry/test/registry.spec.ts [hunk 5]
```diff
@@ -700,6 +700,42 @@ describe('docregistry/registry', () => {
...
+        registry.addFileType({
+          name: 'test_ipynb',
+          contentType: 'notebook',
+          pattern: '^test.ipynb$'
+        });
+        const regularDirectoryFt = registry.getFileTypeForModel({
+          name: 'foo.ipynb',
+          type: 'notebook'
+        });
+        expect(regularDirectoryFt.name).toBe('notebook');
```
Comment (krassowski) [inline packages/docregistry/test/registry.spec.ts:731]:
```suggestion
        const regularNotebookFt = registry.getFileTypeForModel({
          name: 'foo.ipynb',
          type: 'notebook'
        });
        expect(regularNotebookFt.name).toBe('notebook');
```

**Review by krassowski — COMMENTED** [review]
Thank you! Just some minor suggestions on naming things in the tests, otherwise this looks good to me. Do not worry about visual regressions/link test failing (though that should be fixed now) - I will restart these as needed.

**Comment by linlol** [comment]
hello @krassowski, thanks for pointing out the typo, those typos shall be resolved

**Comment by linlol** [comment]
well @krassowski pls help retry

**Review by krassowski — APPROVED** [review]
Thank you @linlol 

CC @mwouts and @mahendrapaipuri for awareness; I think it does not impact juyptext because it restricts its custom notebook types to files [with a different extension than `.ipynb`](https://github.com/mwouts/jupytext/blob/1ef1dea7d1d63e418ecc093025510e7b941ade9c/jupyterlab/packages/jupyterlab-jupytext-extension/src/registry.ts#L35-L67) but it might allow you to overwrite factory for `ipynb` if you need it.

**Comment by mahendrapaipuri** [comment]
Cheers @krassowski for the heads up. Will look into it!!

**Comment by linlol** [comment]
thanks everyone

**Comment by krassowski** [comment]
It is a problem with the tests running in parallel and the kernels not being mocked but pooled from the same server which is used to run multiple tests. #15845 might alleviate it a bit.

**Comment by linlol** [comment]
Thanks @krassowski for your help. As a seven-year jupyter user, it's my great honour to contribute it

**Pull request merged**